### ISSUE TITLE
preprocess2 is not compatible with chunk=to-content

### ISSUE THREAD

**Issue by 88ning** [dita-ot/dita-ot#3514]
## Expected Behavior
If you extend the normalize plugin with a plugin that uses map-first preprocessing, then the normalized output will still support chunking.

## Actual Behavior
Topicrefs with the chunked topics appear as broken links and look like this:

```xml
<topicref href="e9a94ae80eb9ac5e81b9c6b56facbc52bbced403.dita#reference_xlm_4lf_qn">
```

## Possible Solution
Could chunking add that root element ID to the href and this messes up one of the preprocess steps that resolve the hashkeys?

## Steps to Reproduce

1. Create an extension to the normalize plugin.
2. In that new plugin use the preprocess2 target instead of preprocess.
3. Create a ditamap with a chunked topic.
4. Publish the new normalized output.

* DITA-OT version: 3.5
* Operating system and version:
macOS
* How did you run DITA-OT?
  dita command, oxygen
* Transformation type:
  dita normalize

**Comment by ekimbernow** [comment]
I have confirmed that through OT 4.0.1, chunk="to-content" produces incorrect HTML using HTML5 modified to use preprocess2.

The issue appears to be that the chunked result has incorrect root-level PIs for setting the paths. Using preprocess2, I see this in the chunked topic's temp version:

```
<?xml version="1.0" encoding="UTF-8"?><?workdir /Users/eliot.kimber/temp/topic-html5?>
<?workdir-uri file:/Users/eliot.kimber/temp/topic-html5/?>
<?path2rootmap-uri ./?>
<?path2project?><?path2project-uri ./?>
<dita xmlns:ditaarch="http://dita.oasis-open.org/architecture/2005/" ditaarch:DITAArchVersion="1.3">
  <?workdir /Users/eliot.kimber/temp/topic-html5?>
  <?workdir-uri file:/Users/eliot.kimber/temp/topic-html5/?>
  <?path2project ../../?>
  <?path2project-uri ../../?>
  <?path2rootmap-uri ../../?>
  <concept xmlns:dita-ot="http://dita-ot.sourceforge.net/ns/201007/dita-ot" id="chunked-1"
```
Note that the PIs before the <dita> element reflect the root dir of the content (same as the root map), while the ones for the <concept> are correct.

But with preprocess I see this:

```
<?xml version="1.0" encoding="UTF-8"?>
<?workdir /Users/eliot.kimber/temp/topic-html5/content/topics1?>
<?workdir-uri file:/Users/eliot.kimber/temp/topic-html5/content/topics1/?>
<?path2rootmap-uri ./?>
<?path2project ../../?>
<?path2project-uri ../../?>
<dita xmlns:ditaarch="http://dita.oasis-open.org/architecture/2005/" ditaarch:DITAArchVersion="1.3">
  <?workdir /Users/eliot.kimber/temp/topic-html5/content/topics1?>
  <?workdir-uri file:/Users/eliot.kimber/temp/topic-html5/content/topics1/?>
  <?path2project ../../?>
  <?path2project-uri ../../?>
  <?path2rootmap-uri ../../?>
```
Both sets of PIs are correct.

So there must be a bug in the preprocess2 chunk processing.

I will investigate as I need a fix for this ASAP.

[map-with-chunking.zip](https://github.com/dita-ot/dita-ot/files/10484361/map-with-chunking.zip)

**Comment by ekimbernow** [comment]
Digging into the code, as far as I can tell, both preprocess and preprocess2 call the same Java code, ChunkTopicParser, in the same way.

The code that generates the initial path2rootmap PI is:

```
                final File path2rootmap = toFile(getRelativePath(outputFileName, job.getInputMap())).getParentFile();
                writeProcessingInstruction(ditaFileOutput, PI_PATH2ROOTMAP_TARGET_URI, path2rootmap == null ? "./" : toURI(path2rootmap).toString());
```

So the only thing that makes sense is that `outputFileName` is just a name in preprocess2 and the path to the result chunk in preprocess.

 Question is--how could that be? There must be some property setting earlier in the processing chain that is not matching between preprocess and preprocess2 but I have not yet identified it.

**Comment by ekimbernow** [comment]
The difference I'm seeing between preprocess and preprocess2 is that during chunk processing for preprocess2 all the files are generated files in the top-level directory, so they are all in the same directory as the root map:
```
topic-chunk:
    [chunk] Processing file:/Users/eliot.kimber/temp/topic-html5/b5ef8fe69723339d382935fcaaca6bf087c1feaf.ditamap
    [chunk] processCombineChunk(): Before call to chunkParser.write(), currentFile="file:/Users/eliot.kimber/temp/topic-html5/b5ef8fe69723339d382935fcaaca6bf087c1feaf.ditamap"
    [chunk] *** ChunkTopicParser.write(), currentFile="file:/Users/eliot.kimber/temp/topic-html5/b5ef8fe69723339d382935fcaaca6bf087c1feaf.ditamap"
    [chunk] *** ChunkTopicParser.processChunk(): Starting: outputFile="null"
    [chunk]    ChunkTopicParser.processChunk(), parseFilePath="2300667368354b521f958e8483841c3a69719b4c.dita"
    [chunk]    ChunkTopicParser.processChunk(), No 'select-topic' in @chunk, hrefValue="2300667368354b521f958e8483841c3a69719b4c.dita"
    [chunk]    ChunkTopicParser.processChunk(), No 'select-topic' in @chunk, outputFileName="file:/Users/eliot.kimber/temp/topic-html5/2300667368354b521f958e8483841c3a69719b4c.dita"
    [chunk]    ChunkTopicParser.processChunk(), outputFileName is in job store, generating a new name:
    [chunk]    ChunkTopicParser.processChunk(),   Generated outputFileName="file:/Users/eliot.kimber/temp/topic-html5/f19c4f69bb874f13e34320412018069f795a38be.dita"
    [chunk]    ChunkTopicParser.processChunk(), path="file:/Users/eliot.kimber/temp/topic-html5/2300667368354b521f958e8483841c3a69719b4c.dita"
    [chunk]    ChunkTopicParser.processChunk(), newpath="file:/Users/eliot.kimber/temp/topic-html5/f19c4f69bb874f13e34320412018069f795a38be.dita#chunked-1"
    [chunk] Processing file:/Users/eliot.kimber/temp/topic-html5/2300667368354b521f958e8483841c3a69719b4c.dita
    [chunk] processChunk(): in topicref.hasChildNodes(): outputFileName="file:/Users/eliot.kimber/temp/topic-html5/f19c4f69bb874f13e34320412018069f795a38be.dita"
    [chunk] *** ChunkTopicParser.processChunk(): Starting: outputFile="file:/Users/eliot.kimber/temp/topic-html5/f19c4f69bb874f13e34320412018069f795a38be.dita"
    [chunk]    ChunkTopicParser.processChunk(), parseFilePath="12914df527b969677440013f274c316b6e79ebce.dita"
    [chunk]    ChunkTopicParser.processChunk(), path="file:/Users/eliot.kimber/temp/topic-html5/12914df527b969677440013f274c316b6e79ebce.dita"
    [chunk]    ChunkTopicParser.processChunk(), newpath="file:/Users/eliot.kimber/temp/topic-html5/f19c4f69bb874f13e34320412018069f795a38be.dita#chunked-child-01"
    [chunk] Processing file:/Users/eliot.kimber/temp/topic-html5/12914df527b969677440013f274c316b6e79ebce.dita
    [chunk] Writing file:/Users/eliot.kimber/temp/topic-html5/f19c4f69bb874f13e34320412018069f795a38be.dita
    [chunk] writeToContentChunk(): outputFileName="file:/Users/eliot.kimber/temp/topic-html5/f19c4f69bb874f13e34320412018069f795a38be.dita"
    [chunk] writeToContentChunk(): path2rootmap="null"
    [chunk] Process file:/Users/eliot.kimber/temp/topic-html5/b5ef8fe69723339d382935fcaaca6bf087c1feaf.ditamap
    [chunk] Process file:/Users/eliot.kimber/temp/topic-html5/31f8dc4cc97d7bae82500267d1322757951e4867.dita
    [chunk] Process file:/Users/eliot.kimber/temp/topic-html5/2300667368354b521f958e8483841c3a69719b4c.dita
    [chunk] Process file:/Users/eliot.kimber/temp/topic-html5/2002c3348b7941a2c449ef7fabfac2f99d2be270.dita
   [chunk2] Processing file:/Users/eliot.kimber/temp/topic-html5/b5ef8fe69723339d382935fcaaca6bf087c1feaf.ditamap
```
And only later rewritten to their origin relative locations. But by that time the chunk has already been written with the PIs constructed using the temporary files, not the eventual files.

The same output for preprocess is:

```
chunk:
    [chunk] Processing file:/Users/eliot.kimber/temp/topic-html5/map-with-chunking.ditamap
    [chunk] processCombineChunk(): Before call to chunkParser.write(), currentFile="file:/Users/eliot.kimber/temp/topic-html5/map-with-chunking.ditamap"
    [chunk] *** ChunkTopicParser.write(), currentFile="file:/Users/eliot.kimber/temp/topic-html5/map-with-chunking.ditamap"
    [chunk] *** ChunkTopicParser.processChunk(): Starting: outputFile="null"
    [chunk]    ChunkTopicParser.processChunk(), parseFilePath="content/topics1/chunked-1.dita"
    [chunk]    ChunkTopicParser.processChunk(), No 'select-topic' in @chunk, hrefValue="content/topics1/chunked-1.dita"
    [chunk]    ChunkTopicParser.processChunk(), No 'select-topic' in @chunk, outputFileName="file:/Users/eliot.kimber/temp/topic-html5/content/topics1/chunked-1.dita"
    [chunk]    ChunkTopicParser.processChunk(), outputFileName is in job store, generating a new name:
    [chunk]    ChunkTopicParser.processChunk(),   Generated outputFileName="file:/Users/eliot.kimber/temp/topic-html5/Chunk1284895233.dita"
    [chunk]    ChunkTopicParser.processChunk(), path="file:/Users/eliot.kimber/temp/topic-html5/content/topics1/chunked-1.dita"
    [chunk]    ChunkTopicParser.processChunk(), newpath="file:/Users/eliot.kimber/temp/topic-html5/Chunk1284895233.dita#chunked-1"
    [chunk] Processing file:/Users/eliot.kimber/temp/topic-html5/content/topics1/chunked-1.dita
    [chunk] processChunk(): in topicref.hasChildNodes(): outputFileName="file:/Users/eliot.kimber/temp/topic-html5/Chunk1284895233.dita"
    [chunk] *** ChunkTopicParser.processChunk(): Starting: outputFile="file:/Users/eliot.kimber/temp/topic-html5/Chunk1284895233.dita"
    [chunk]    ChunkTopicParser.processChunk(), parseFilePath="content/topics2/chunked-child-01.dita"
    [chunk]    ChunkTopicParser.processChunk(), path="file:/Users/eliot.kimber/temp/topic-html5/content/topics2/chunked-child-01.dita"
    [chunk]    ChunkTopicParser.processChunk(), newpath="file:/Users/eliot.kimber/temp/topic-html5/Chunk1284895233.dita#chunked-child-01"
    [chunk] Processing file:/Users/eliot.kimber/temp/topic-html5/content/topics2/chunked-child-01.dita
    [chunk] Writing file:/Users/eliot.kimber/temp/topic-html5/Chunk1284895233.dita
    [chunk] writeToContentChunk(): outputFileName="file:/Users/eliot.kimber/temp/topic-html5/Chunk1284895233.dita"
    [chunk] writeToContentChunk(): path2rootmap="null"
    [chunk] Process file:/Users/eliot.kimber/temp/topic-html5/content/topics1/not-chunked-1.dita
    [chunk] Process file:/Users/eliot.kimber/temp/topic-html5/othercontent/subfolder/map-with-chunking.dita
    [chunk] Process file:/Users/eliot.kimber/temp/topic-html5/map-with-chunking.ditamap
    [chunk] Process file:/Users/eliot.kimber/temp/topic-html5/content/topics1/chunked-1.dita
   [chunk2] Processing file:/Users/eliot.kimber/temp/topic-html5/map-with-chunking.ditamap
```

I don't see an obvious way to resolve this for preprocess2.

**Comment by jomjohn** [comment]
Hi @ekimbernow, I am testing preprocess2 with HTML5 w/dita-ot 3.7.3 and I am unable to reproduce missing images using your sample [map-with-chunking.zip](https://github.com/dita-ot/dita-ot/files/10484361/map-with-chunking.zip) above.

I see that you are using a keydef for the image. I find when the keydef is **not** used and the topic uses a direct path to the image it is then broken.

Can you please confirm my last statement? thanks

**Comment by jomjohn** [comment]
Hi @ekimbernow, I am testing preprocess2 with HTML5 w/dita-ot 3.7.3 and I am unable to reproduce missing images using your sample [map-with-chunking.zip](https://github.com/dita-ot/dita-ot/files/10484361/map-with-chunking.zip) above. Can you point out which html files have broken references?

I am able to produce the missing image issue when chunk=to-content is defined at the root map level as in the example from #4042 [chunk.zip](https://github.com/dita-ot/dita-ot/files/9972528/chunk.zip)

**Comment by ekimbernow** [comment]
Rerunning my test now--the issue isn't the images (at least not with keys), it's relative CSS in the generated HTML is not relative to the location of the HTML but of the root directory:

```
<link rel="stylesheet" type="text/css" href="./oxygen-webhelp/app/commons.css?buildId=2022101006"/>
<link rel="stylesheet" type="text/css" href="./oxygen-webhelp/app/topic.css?buildId=2022101006"/>
```

The HTML is at `webhelp-responsive/map-with-chunking/content/topics1/chunked-1.html`

In a non-chunked topic the css links are:

```
<link rel="stylesheet" type="text/css" href="../../oxygen-webhelp/app/commons.css?buildId=2022101006"/>
<link rel="stylesheet" type="text/css" href="../../oxygen-webhelp/app/topic.css?buildId=2022101006"/>
```
Note the correct relative path in this case.

The content we were publishing when we first noticed the issue did not use keys for images, so the relative image references were not working.

We have now migrated all our content to use keys for images, so that would not be an issue. But having the CSS references not work is a hard failure as we use Oxygen responsive web help for one of our deliverables (along with Zoomin, which I think is insensitive to this particular issue for CSS because it's a React responsive site that just grabs the juicy center of the HTML files.
